lib/resty/auto-ssl/jobs/renewal.lua

```diff
@@ -79,8 +79,43 @@ local function renew_check_cert(auto_ssl_instance, storage, domain)
     cert = {}
   end
 
-  -- Attempt to retrieve expiry date from storage. If it is not found try renewal.
-  -- If expiry date is found, we attempt renewal if it's within 30 days.
+  if not cert["fullchain_pem"] then
+    ngx.log(ngx.ERR, "auto-ssl: attempting to renew certificate for domain without certificates in storage: ", domain)
+    renew_check_cert_unlock(domain, storage, local_lock, distributed_lock_value)
+    return
+  end
+
+  -- If we don't have an expiry date yet, try to update the stored cert
+  -- with a date based on backup timestamps
+  if not cert["expiry"] then
+    local keys, err = storage.adapter:keys_with_prefix(domain .. ":")
```

Comment:
Rather basing the missing `expiry` times on the storage key's timestamps, what about instead using `run_command` to run the same shell commands we use to fetch this directly from the certificate data using `openssl`? https://github.com/GUI/lua-resty-auto-ssl/blob/v0.12.0/bin/letsencrypt_hooks#L37 It would require writing the cert data to a temporary file on disk, but since this should only need to happen to upgrade pre-v0.12 data once, this overhead doesn't seem like it would be a big deal.

It seems like this approach would ensure better consistency, since we'd be sure we get the accurate expiration time directly out of the certificate (in case the timestamp it was stored at was different for whatever reason). It would also work, even if we made changes to potentially get rid of the old timestamped certs from storage, as discussed in https://github.com/GUI/lua-resty-auto-ssl/pull/124

Reply:
This idea made sense to me. I went for the timestamps, since it seemed more obtainable (I am still only dabbling in Lua).

Note that getting rid of newly-created backups down the line would not cause an issue here, as all certificates created after af157713ffe80f44bb36d8cf3a9af2c374763b96 already have the `expiry` information set.